internal/batch/slice.go

```diff
@@ -0,0 +1,143 @@
+)
+
+type (
+	SlicedRecord struct {
+		arrow.Record
+		Bytes       int64 // we need this as the util.TotalRecordSize will report the full size even for the sliced record
+		bytesPerRow int64
+	}
+)
+
+func (s *SlicedRecord) split(limit *Cap) (add *SlicedRecord, toFlush []arrow.Record, rest *SlicedRecord) {
+	if s == nil {
+		return nil, nil, nil
+	}
+
+	add = s.getAdd(limit)
+	if add != nil {
+		limit.add(add.Bytes, add.NumRows())
+	}
+
+	if s.Record == nil {
+		// all processed
+		return add, nil, nil
+	}
+
+	toFlush = s.getToFlush(limit)
+	if s.Record == nil {
+		// all processed
+		return add, toFlush, nil
+	}
+
+	// set bytes & rows new values
+	limit.set(s.Bytes, s.NumRows())
+	return add, toFlush, s
+}
+
+func (s *SlicedRecord) getAdd(limit *Cap) *SlicedRecord {
+	rowsByBytes := limit.bytes.remainingPerN(s.bytesPerRow)
+	rows := limit.rows.remaining()
+	switch {
+	case rows < 0:
+		rows = rowsByBytes
+	case rows > rowsByBytes && rowsByBytes >= 0:
+		rows = rowsByBytes
+	}
+
+	switch {
+	case rows == 0:
+		return nil
+	case rows < 0, rows >= s.NumRows():
+		// grab the whole record (either no limits or not overflowing)
+		res := *s
+		s.Bytes = 0
+		s.Record = nil
+		return &res
+	}
+
+	res := SlicedRecord{
+		Record:      s.NewSlice(0, rows),
+		Bytes:       rows * s.bytesPerRow,
+		bytesPerRow: s.bytesPerRow,
+	}
+	s.Record = s.NewSlice(rows, s.NumRows())
+	s.Bytes -= res.Bytes
+	return &res
+}
+
+func (s *SlicedRecord) getToFlush(limit *Cap) []arrow.Record {
+	rowsByBytes := limit.bytes.capPerN(s.bytesPerRow)
+	rows := limit.rows.cap()
+	switch {
+	case rows < 0:
+		rows = rowsByBytes
+	case rows > rowsByBytes && rowsByBytes >= 0:
+		rows = rowsByBytes
+	}
+
+	switch {
+	case rows == 0:
+		// not even a single row fits
+		// we still need to process this, so slice by single row
+		return s.slice()
+	case rows < 0:
+		// as s.Record != nil we know that the limits are there in place & the s.Record.NumRows() > 0
+		panic("should never be here")
+	case rows > s.NumRows():
+		// no need to flush anything, as the amount of rows isn't enough to grant this
+		return nil
+	}
+
+	flush := make([]arrow.Record, 0, s.NumRows()/rows)
+	offset := int64(0)
+	for offset+rows <= s.NumRows() {
+		flush = append(flush, s.NewSlice(offset, offset+rows))
+		offset += rows
+	}
+	if offset == s.NumRows() {
+		// we processed everything for flush
+		s.Record = nil
+		s.Bytes = 0
+		return flush
+	}
+
+	// set record to the remainder
+	s.Record = s.NewSlice(offset, s.NumRows())
+	s.Bytes = s.NumRows() * s.bytesPerRow
+
+	return flush
+}
+
+func (s *SlicedRecord) slice() []arrow.Record {
+	res := make([]arrow.Record, s.NumRows())
+	for i := int64(0); i < s.NumRows(); i++ {
+		res[i] = s.NewSlice(i, i+1)
+	}
+	return res
+}
+
+func newSlicedRecord(r arrow.Record) *SlicedRecord {
+	if r.NumRows() == 0 {
+		return nil
+	}
+	res := SlicedRecord{
+		Record: r,
+		Bytes:  util.TotalRecordSize(r),
+	}
+	res.bytesPerRow = res.Bytes / r.NumRows()
```

Comment:
Still going through this PR, but is it true that we're assuming each row has the same number of bytes?
Can't we get the number of bytes per row?

Reply:
we can't get amount of bytes per row (we could get it for fixed-size ones, but for variable-sized ones it'll create too much overhead)

Reply:
Doesn't this mean we can't reliably enforce the bytes limits? Let me try this out on a destination that needs the bytes limit to be a hard one, and not a soft one.
Maybe we can find a solution for those type of destinations

Reply:
We can't make a 100% correct implementation without making a huge toll on CPU, as currently even the sliced array will still have all buffers of the original mentioned & thus we would have the total bytes for the array equal to the value for slice of even a single row.
While we don't actually impose the "technically correct" limit in here, having this PR at least enables slicing huge records into smaller chunks to try to accommodate for the destinations that really require that.

In the current mainline the huge record with lots of rows that is x10 times more than the destination limit will be flushed all at once as a single write. This change addresses such issues.

Reply:
> In the current mainline the huge record with lots of rows

But we only support records with multiple rows in a few DB plugins (and not as the default setting), and the intention with this change is to have all sources send multi row records by default.

We should first check how likely it is that we breach the limit, and the impact on destinations (not all destinations need a hard limit).